File: LeetCode/reverse_linked_list.cpp
/*
Problem Link:
https://leetcode.com/problems/reverse-linked-list/description/
*/


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(!head) return head;
        ListNode *tempHead = NULL;
        ListNode *temp = head;
        while(head){
            if(!tempHead){
                tempHead = head;
                head = head->next;
                tempHead->next =NULL;
                temp = head;
            }
            else{
                temp = temp->next;
                head->next = tempHead;
                tempHead= head;
                head = temp;
            }
        }
        return tempHead;
    }
};

File: LeetCode/reverse_linked_list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if(head == null) return head;
        ListNode nextNode = head.next;
        ListNode newHead = head;
        ListNode currentNode = head.next;
        newHead.next = null;
        while(nextNode != null){
            nextNode = currentNode.next;
            currentNode.next = newHead;
            newHead = currentNode;
            currentNode = nextNode;
        }
        return newHead;
    }
}